com.unity.perception/Runtime/GroundTruth/Labelers/Visualization/HUDPanel.cs

```diff
@@ -13,119 +14,125 @@ namespace UnityEngine.Perception.GroundTruth
+            GUI.skin.label.margin = new RectOffset(0, 0, 1, 1);
+            GUI.skin.box.padding = new RectOffset(5, 5, 5, 5);
+            GUI.skin.toggle.margin = new RectOffset(0, 0, 0, 0);
+            GUI.skin.horizontalSlider.margin = new RectOffset(0, 0, 0, 0);
+            guiStylesInitialized = true;
+        }
 
-            if (!entries.ContainsKey(key))
+        int EntriesCount()
+        {
+            return entries.Count + entries.Sum(entry => entry.Value.Count);
+        }
+
+        internal void onDrawGUI()
+        {
+            if (entries.Count == 0) return;
+
+            if (!guiStylesInitialized) SetUpGUIStyles();
+
+            GUI.depth = 0; // Draw HUD objects on the top of other UI objects
+
+            var height = Math.Min(LineHeight * EntriesCount(), Screen.height * 0.5f - YPadding * 2);
+            var xPos = Screen.width - BoxWidth - XPadding;
+            var yPos = Screen.height - height - YPadding;
+
+            GUILayout.BeginArea(new Rect(xPos, yPos, BoxWidth, height), GUI.skin.box);
+
+            scrollPosition = GUILayout.BeginScrollView(scrollPosition);
+
+            bool firstTime = true;
+            foreach (var labeler in entries.Keys)
             {
-                if (orphans.Any())
+                if (!firstTime) GUILayout.Space(YLineSpacing);
+                firstTime = false;
+                GUILayout.Label(labeler);
+                foreach (var entry in entries[labeler])
                 {
-                    val = (true, orphans.Pop());
-                    val.Item2.gameObject.SetActive(true);
+                    GUILayout.BeginHorizontal();
+                    GUILayout.Space(5);
+                    GUILayout.Label(entry.Key);
+                    GUILayout.FlexibleSpace();
+                    GUILayout.Label(entry.Value);
+                    GUILayout.EndHorizontal();
                 }
-                else
-                {
-                    val = (true, GameObject.Instantiate(Resources.Load<GameObject>("KeyValuePanel")).GetComponent<KeyValuePanel>());
-                    val.Item2.transform.SetParent(contentPanel.transform, false);
-                }
-
-                val.Item2.SetKey(key);
-            }
-            else
-            {
-                val = entries[key];
-                val.Item1 = true;
             }
 
-            val.Item2.SetValue(value);
-            entries[key] = val;
+            GUILayout.EndScrollView();
+            GUILayout.EndArea();
         }
 
         /// <summary>
         /// Removes the key value pair from the HUD
         /// </summary>
+        /// <param name="labeler">The labeler that requested the removal</param>
         /// <param name="key">The key of the entry to remove</param>
-        public void RemoveEntry(string key)
+        public void RemoveEntry(CameraLabeler labeler, string key)
         {
-            if (entries.ContainsKey(key))
+            if (entries.ContainsKey(labeler.GetType().Name))
```

Comment:
It is possible to have multiple labelers of the same type. We'll need to support that somehow.

Reply:
It would be great to add a test for this case as well.